go/services/reports/pkg/incident_correlation/utils.go

```diff
@@ -0,0 +1,100 @@
+package incidentcorrelation
+
+import (
+	"slices"
+
+	"github.com/Magpie-Monitor/magpie-monitor/services/reports/pkg/repositories"
+)
+
+func ConvertConcreteIncidentArrayIntoIncidents[T repositories.Incident](incidents []T) []repositories.Incident {
+	convertedIncidents := make([]repositories.Incident, 0, len(incidents))
+
+	for _, incident := range incidents {
+		convertedIncidents = append(convertedIncidents, incident)
+	}
+
+	return convertedIncidents
+}
+
+// Removes merged incidents from report and returns new incident as a result
+func mergeApplicationIncidents(group IncidentMergeGroup, applicationIncidents *repositories.ApplicationReport) *repositories.ApplicationIncident {
+	var sourcesToBeMerged []repositories.ApplicationIncidentSource
+	var filteredIncidents []*repositories.ApplicationIncident
+
+	for _, incident := range applicationIncidents.Incidents {
+		if !slices.Contains(group.IncidentIds, incident.Id) {
+			filteredIncidents = append(filteredIncidents, incident)
+			continue
+		}
+
+		sourcesToBeMerged = append(sourcesToBeMerged, incident.Sources...)
+	}
+
+	newIncident := &repositories.ApplicationIncident{
+		Title:           group.Title,
+		Urgency:         group.Urgency,
+		Summary:         group.Summary,
+		Recommendation:  group.Recommendation,
+		Category:        group.Category,
+		CustomPrompt:    applicationIncidents.CustomPrompt,
+		Accuracy:        applicationIncidents.Accuracy,
+		ApplicationName: applicationIncidents.ApplicationName,
+		ClusterId:       applicationIncidents.Incidents[0].ClusterId,
```

Comment:
What if Incidents is of length 0?

Reply:
This is not possible, as incident group will always have at least one element

Reply:
But it's a good point to add a check before with more descriptive error. I will add this in this PR.